storey/aggregations.py

```diff
@@ -144,41 +199,131 @@ class AggregateStore:
     def __init__(self, aggregates):
         self.cache = {}
         self.aggregates = aggregates
+        self._storage = None
+        self._table_path = None
+        self.schema = None
+        self.read_only = False
 
     def __iter__(self):
         return iter(self.cache.items())
 
-    def aggregate(self, key, data, timestamp):
+    async def aggregate(self, key, data, timestamp):
+        if not self.schema:
+            await self.get_or_save_schema()
+
         if isinstance(timestamp, datetime):
             timestamp = timestamp.timestamp() * 1000
 
-        if key not in self.cache:
-            self.cache[key] = AggregatedStoreElement(key, self.aggregates, timestamp)
+        key_to_aggregate = await self._get_or_load_key(key, timestamp)
+        key_to_aggregate.aggregate(data, timestamp)
 
-        self.cache[key].aggregate(data, timestamp)
+    async def get_features(self, key, timestamp):
+        if not self.schema:
+            await self.get_or_save_schema()
 
-    def get_features(self, key, timestamp):
         if isinstance(timestamp, datetime):
             timestamp = timestamp.timestamp() * 1000
 
-        return self.cache[key].get_features(timestamp)
+        relevant_key = await self._get_or_load_key(key, timestamp)
+        return relevant_key.get_features(timestamp)
+
+    async def _get_or_load_key(self, key, timestamp=None):
+        if self.read_only or key not in self.cache:
+            # Try load from the store, and create a new one only if the key really is new
+            initial_data = await self._storage._load_key(self._table_path, key)
+            self.cache[key] = AggregatedStoreElement(key, self.aggregates, timestamp, initial_data)
+
+        return self.cache[key]
 
     def get_keys(self):
         return self.cache.keys()
 
+    async def get_or_save_schema(self):
+        self.schema = await self._storage._load_schema(self._table_path)
+        should_update = not self.schema
+        if self.schema:
+            should_update = self._validate_schema_fit_aggregations(self.schema)
+
+        if should_update:
+            self.schema = await self._save_schema()
+
+    async def _save_schema(self):
+        schema = self._aggregates_to_schema()
+        if self.schema:
+            schema = self.merge_schemas(self.schema, schema)
+
+        await self._storage._save_schema(self._table_path, schema)
+        return schema
+
+    def merge_schemas(self, old, new):
```

Comment:
Should be private. Do we test this scenario?

Reply:
did manually, ill add more tests